File: tugas_mahasiswa.py
# Inisialisasi kapasitas 10 dengan nilai None (sebagai pengganti null di Java)
kapasitas = 10
data_mahasiswa = [[None, None] for _ in range(kapasitas)]
count = 0

def show_data():
    if count == 0:
        print("\nArray Kosong (Tidak ada data mahasiswa).")
    else:
        print("\n--- LIST DATA MAHASISWA ---")
        for i in range(count):
            print(f"Index [{i}] -> Nama: {data_mahasiswa[i][0]} | NIM: {data_mahasiswa[i][1]}")

def insert_at_pos(posisi):
    global count
    if count >= kapasitas:
        print("Gagal: Kapasitas penuh!")
        return
    if posisi < 0 or posisi > count:
        print("Gagal: Posisi tidak valid!")
        return

    # Input data
    nama = input("Masukkan Nama: ")
    nim = input("Masukkan NIM : ")

    # Shifting (Geser ke kanan)
    # Di Python kita bisa pakai slice atau loop mundur
    for i in range(count, posisi, -1):
        data_mahasiswa[i] = data_mahasiswa[i-1]

    data_mahasiswa[posisi] = [nama, nim]
    count += 1
    print(f"Data berhasil ditambahkan di index {posisi}")

def delete_at_pos(posisi):
    global count
    if count == 0:
        print("Gagal: Data kosong!")
        return
    if posisi < 0 or posisi >= count:
        print("Gagal: Index tidak ditemukan!")
        return

    # Shifting (Geser ke kiri)
    for i in range(posisi, count - 1):
        data_mahasiswa[i] = data_mahasiswa[i+1]
    
    data_mahasiswa[count-1] = [None, None]
    count -= 1
    print(f"Data di index {posisi} berhasil dihapus.")

def delete_first_occurrence(target):
    for i in range(count):
        if data_mahasiswa[i][0].lower() == target.lower():
            delete_at_pos(i)
            return
    print(f"Data dengan nama '{target}' tidak ditemukan.")

# Main Program
while True:
    print("\n======= MENU DATA MAHASISWA (PYTHON) =======")
    print(f"Data Terisi: {count} / {kapasitas}")
    print("--------------------------------------------")
    print("1. Insert at Beginning")
    print("2. Insert at Given Position")
    print("3. Insert at End")
    print("4. Delete from Beginning")
    print("5. Delete Given Position")
    print("6. Delete from End")
    print("7. Delete First Occurrence")
    print("8. Show Data")
    print("9. Exit")
    
    pilihan = input("Pilih menu (1-9): ")

    if pilihan == '1':
        insert_at_pos(0)
    elif pilihan == '2':
        pos = int(input(f"Masukkan posisi index (0-{count}): "))
        insert_at_pos(pos)
    elif pilihan == '3':
        insert_at_pos(count)
    elif pilihan == '4':
        delete_at_pos(0)
    elif pilihan == '5':
        pos = int(input("Masukkan index yang dihapus: "))
        delete_at_pos(pos)
    elif pilihan == '6':
        delete_at_pos(count - 1)
    elif pilihan == '7':
        target = input("Masukkan nama yang dihapus: ")
        delete_first_occurrence(target)
    elif pilihan == '8':
        show_data()
    elif pilihan == '9':
        print("Program Selesai Yey!")
        break
    else:
        print("Pilihan tidak valid!")

File: TugasMahasiswa.java
// Nama : Mahardika Ahmad
// NIM  : 24106050077

import java.util.Scanner;

public class TugasMahasiswa {
    // Kapasitas array tetap 10 sesuai aturan
    static int kapasitas = 10;
    // Kolom 0 untuk Nama, Kolom 1 untuk NIM
    static String[][] dataMahasiswa = new String[kapasitas][2];
    // Variabel count untuk melacak jumlah data terisi
    static int count = 0;

    public static void main(String[] args) {
        Scanner input = new Scanner(System.in);
        int pilihan;

        do {
            System.out.println("\n======= MENU DATA MAHASISWA =======");
            System.out.println("Data Terisi: " + count + " / " + kapasitas);
            System.out.println("-----------------------------------");
            System.out.println("1. Insert at Beginning");
            System.out.println("2. Insert at Given Position");
            System.out.println("3. Insert at End");
            System.out.println("4. Delete from Beginning");
            System.out.println("5. Delete Given Position");
            System.out.println("6. Delete from End");
            System.out.println("7. Delete First Occurrence (By Nama)");
            System.out.println("8. Show Data");
            System.out.println("9. Exit");
            System.out.print("Pilih menu (1-9): ");
            pilihan = input.nextInt();
            input.nextLine(); // Membersihkan buffer enter

            switch (pilihan) {
                case 1:
                    insertAtPos(0, input);
                    break;
                case 2:
                    System.out.print("Masukkan posisi index (0-" + count + "): ");
                    int pos = input.nextInt();
                    input.nextLine();
                    insertAtPos(pos, input);
                    break;
                case 3:
                    insertAtPos(count, input);
                    break;
                case 4:
                    deleteAtPos(0);
                    break;
                case 5:
                    System.out.print("Masukkan posisi index yang dihapus: ");
                    int delPos = input.nextInt();
                    deleteAtPos(delPos);
                    break;
                case 6:
                    deleteAtPos(count - 1);
                    break;
                case 7:
                    System.out.print("Masukkan Nama yang ingin dihapus: ");
                    String namaCari = input.nextLine();
                    deleteFirstOccurrence(namaCari);
                    break;
                case 8:
                    showData();
                    break;
                case 9:
                    System.out.println("Program selesai. Sampai jumpa!");
                    break;
                default:
                    System.out.println("Pilihan tidak valid!");
            }
        } while (pilihan != 9);

        input.close();
    }

    // --- FUNGSI INSERT ---
    static void insertAtPos(int posisi, Scanner in) {
        if (count >= kapasitas) {
            System.out.println("Gagal: Kapasitas array sudah penuh!");
            return;
        }
        if (posisi < 0 || posisi > count) {
            System.out.println("Gagal: Posisi di luar jangkauan!");
            return;
        }

        // Geser data ke kanan untuk memberi ruang di 'posisi'
        for (int i = count; i > posisi; i--) {
            dataMahasiswa[i][0] = dataMahasiswa[i - 1][0];
            dataMahasiswa[i][1] = dataMahasiswa[i - 1][1];
        }

        System.out.print("Masukkan Nama: ");
        dataMahasiswa[posisi][0] = in.nextLine();
        System.out.print("Masukkan NIM : ");
        dataMahasiswa[posisi][1] = in.nextLine();

        count++; // NB 1: count ditambah 1
        System.out.println("Data berhasil ditambahkan di index " + posisi);
    }

    // --- FUNGSI DELETE BERDASARKAN POSISI ---
    static void deleteAtPos(int posisi) {
        if (count == 0) {
            System.out.println("Gagal: Data masih kosong!");
            return;
        }
        if (posisi < 0 || posisi >= count) {
            System.out.println("Gagal: Index tidak ditemukan!");
            return;
        }

        // Geser data ke kiri untuk menimpa data yang dihapus
        for (int i = posisi; i < count - 1; i++) {
            dataMahasiswa[i][0] = dataMahasiswa[i + 1][0];
            dataMahasiswa[i][1] = dataMahasiswa[i + 1][1];
        }

        // Bersihkan bekas data terakhir
        dataMahasiswa[count - 1][0] = null;
        dataMahasiswa[count - 1][1] = null;

        count--; // NB 2: count dikurang 1
        System.out.println("Data pada index " + posisi + " berhasil dihapus.");
    }

    // --- FUNGSI DELETE FIRST OCCURRENCE ---
    static void deleteFirstOccurrence(String target) {
        for (int i = 0; i < count; i++) {
            // Membandingkan teks tanpa peduli besar/kecil huruf
            if (dataMahasiswa[i][0].equalsIgnoreCase(target)) {
                deleteAtPos(i);
                return; // Berhenti setelah menemukan yang pertama
            }
        }
        System.out.println("Data dengan nama '" + target + "' tidak ditemukan.");
    }

    // --- FUNGSI TAMPILKAN DATA ---
    static void showData() {
        if (count == 0) {
            System.out.println("Array Kosong (Tidak ada data mahasiswa).");
        } else {
            System.out.println("\n--- LIST DATA MAHASISWA ---");
            for (int i = 0; i < count; i++) {
                System.out.println("Index [" + i + "] -> Nama: " + dataMahasiswa[i][0] + " | NIM: " + dataMahasiswa[i][1]);
            }
        }
    }
}